### PR TITLE
PLAT-24455 support multi clip captions copy

### PR DESCRIPTION
**Pull request by salma2222**

### CODE REVIEW

File: plugins/content/caption/base/batch/CopyCaptions/KAsyncCopyCaptions.class.php [hunk 1]
```diff
@@ -83,34 +96,54 @@ private function getAllCaptionAsset($entryId)
...
-
-	private function getUnsupportedFormats()
+	protected function getUnsupportedFormats()
 	{
 		$unsupportedFormats = array (CaptionType::CAP, CaptionType::SCC);
 		return $unsupportedFormats;
 	}
 
-	private function cloneCaption($targetEntryId, $originalCaptionAsset)
+	protected function cloneCaptionAsset($targetEntryId, $clonedCaptionAsset)
```
Comment (inbal-ben-david) [inline plugins/content/caption/base/batch/CopyCaptions/KAsyncCopyCaptions.class.php:127]: why the name was changed to cloned? i think it was clearer with previous name

File: plugins/content/caption/base/batch/CopyCaptions/KAsyncCopyCaptions.class.php [hunk 2]
```diff
@@ -189,29 +216,124 @@ private function getCaptionContent($captionAssetId)
...
 	 * @throws kApplicativeException
 	 */
-	private function copyFromClipToDestination(KalturaCopyCaptionsJobData $data, $clipDescriptionArray)
+	protected function copyFromSingleSourceToDestination(KalturaCopyCaptionsJobData $data)
 	{
 		$errorMsg = '';
-		//currently only one source
-		$originalCaptionAssets = $this->getAllCaptionAsset($clipDescriptionArray[0]->sourceEntryId);
+		$clipDescriptionArray = $data->clipsDescriptionArray;
+		$originalCaptionAssets = $this->listCaptionAssets($clipDescriptionArray[0]->sourceEntryId);
```
Comment (inbal-ben-david) [inline plugins/content/caption/base/batch/CopyCaptions/KAsyncCopyCaptions.class.php:225]: for sure this is not empty?

File: plugins/content/caption/base/lib/kCaptionsContentManager.php [hunk 3]
```diff
@@ -307,14 +307,21 @@ public function createCaptionsFile($content, $clipStartTime, $clipEndTime, $time
...
 						$shouldAddBlockToNewFile = false;
+					}
 				}
 				else
+				{
 					$currentBlock .= $line . kCaptionsContentManager::UNIX_LINE_ENDING;
+				}
 				$line = kCaptionsContentManager::getNextValueFromArray($originalFileContentArray);
 			}
+			$shouldAddBlockToNewFile = $shouldAddBlockToNewFile && $currentBlock && trim($currentBlock) !== '';
```
Comment (inbal-ben-david) [inline plugins/content/caption/base/lib/kCaptionsContentManager.php:324]: why needed?

File: plugins/content/caption/base/lib/kCaptionsContentManager.php [hunk 4]
```diff
@@ -328,4 +335,43 @@ protected function addBlockToNewFile($newFileContent, $currentBlock)
...
+		return implode(kCaptionsContentManager::UNIX_LINE_ENDING, $blockArray);
+	}
+
+	protected static function srtToWebvttTimeLine($srtContentManager, $line)
+	{
+		$matches = array();
+		$timecode_match = preg_match(srtCaptionsContentManager::SRT_TIMECODE_PATTERN, $line, $matches);
+		if ($timecode_match)
+		{
+			$startCaption = $srtContentManager->parseCaptionTime($matches[1]);
```
Comment (inbal-ben-david) [inline plugins/content/caption/base/lib/kCaptionsContentManager.php:370]: whats if some of the matches are missing?

File: plugins/content/caption/base/lib/kCaptionsContentManager.php [hunk 5]
```diff
@@ -328,4 +335,43 @@ protected function addBlockToNewFile($newFileContent, $currentBlock)
...
+			$endCaption = $srtContentManager->parseCaptionTime($matches[2]);
+			$timeLine = kWebVTTGenerator::formatWebVTTTimeStamp($startCaption) . ' --> ' . kWebVTTGenerator::formatWebVTTTimeStamp($endCaption);
+			$line = kCaptionsContentManager::UNIX_LINE_ENDING . $timeLine;
```
Comment (inbal-ben-david) [inline plugins/content/caption/base/lib/kCaptionsContentManager.php:373]: not supporting styling was not done as a limitation?

File: plugins/content/caption/base/lib/kCaptionsContentManager.php (hunk 3 above)
Comment (salma2222) [reply plugins/content/caption/base/lib/kCaptionsContentManager.php:324]: we divide the caption content to blocks, by new line, each block will have identifier (number), times and text
for example:
if we have a caption file with 2 new lines at the end then we will create an empty block (with identifier but no time and no text)

File: plugins/content/caption/base/batch/CopyCaptions/KAsyncCopyCaptions.class.php (hunk 2 above)
Comment (salma2222) [reply plugins/content/caption/base/batch/CopyCaptions/KAsyncCopyCaptions.class.php:243]: no, based on the flow it could be empty, added condition

**Review by inbal-ben-david — APPROVED** [review]

**Pull request merged**